Implement IDL changes for `hidden=until-found`

nt1m opened the pull request:
#### 03b9c3b9eda004e6b85f6a116444856e85185912
<pre>
Implement IDL changes for `hidden=until-found`
<a href="https://bugs.webkit.org/show_bug.cgi?id=290369">https://bugs.webkit.org/show_bug.cgi?id=290369</a>
<a href="https://rdar.apple.com/147940452">rdar://147940452</a>

Reviewed by Anne van Kesteren and Darin Adler.

Implement:
- Add preference
- IDL changes specified at: <a href="https://html.spec.whatwg.org/#htmlelement">https://html.spec.whatwg.org/#htmlelement</a>
- DOM getter/setter algorithms specified at: <a href="https://html.spec.whatwg.org/#dom-hidden">https://html.spec.whatwg.org/#dom-hidden</a>

* LayoutTests/imported/w3c/web-platform-tests/html/editing/the-hidden-attribute/hidden-idl-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/editing/the-hidden-attribute/hidden-until-found-idl-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::hidden const):
(WebCore::HTMLElement::setHidden):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLElement.idl:

Canonical link: <a href="https://commits.webkit.org/292735@main">https://commits.webkit.org/292735@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d6f11f81faf231d2ed7d01f63e754602de6dea7b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96929 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16543 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6737 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102003 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47450 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98974 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16839 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24991 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73852 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31061 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99932 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12695 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87684 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54188 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12453 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5509 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46777 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/89602 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/82542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5590 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104026 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/95550 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23997 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/82900 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24250 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83762 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82295 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/20687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26965 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/4513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17500 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15643 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23959 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29114 "Failed to checkout and rebase branch from PR 43043") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119177 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/23787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27098 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25359 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1001,6 +1001,43 @@ void HTMLElement::setEnterKeyHint(const AtomString& value)
+    WTF::switchOn(*value, [&](String string) {
+        if (document().settings().hiddenUntilFoundEnabled() && equalIgnoringASCIICase(string, "until-found"_s))
+            setAttributeWithoutSynchronization(hiddenAttr, "until-found"_s);
+        else if (string.isEmpty() || string.isNull())
```
```suggestion
        else if (string.isEmpty())
```
Null strings are also empty strings.

annevk replied:
I think isEmpty is inclusive of isNull? (It does seem weird that the empty string would remove it as that's a perfectly fine value, but it's in the spec...)

annevk commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -3469,6 +3469,20 @@ HiddenPageDOMTimerThrottlingEnabled:
+   type: bool
+   status: testable
+   category: html
+   humanReadableName: "hidden=until-found"
```
Don't we already support this? This is only hidden getter/setter support for it right?

Assuming we already support this feature I'd enable this by default as this is a trivial change to the getter/setter. But we should change the name here then to make it more clear this is solely about the getter/setter.

nt1m replied:
I don't think we do? There's some work around the reveal algorithm, which removes the `hidden` attribute when find-in-page reaches it, and emits a `beforematch` event.

darinadler commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -138,6 +138,9 @@ class HTMLElement : public StyledElement {
     void setEnterKeyHint(const AtomString& value);
 
+    std::optional<std::variant<bool, double, String>> hidden() const;
+    void setHidden(const std::optional<std::variant<bool, double, String>>);
```
```suggestion
    void setHidden(const std::optional<std::variant<bool, double, String>>&);
```

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1001,6 +1001,43 @@ void HTMLElement::setEnterKeyHint(const AtomString& value)
+        return;
+    }
+
+    WTF::switchOn(*value, [&](String string) {
```
```suggestion
    WTF::switchOn(*value, [&](const String& string) {
```

nt1m committed:
Implement IDL changes for `hidden=until-found`
https://bugs.webkit.org/show_bug.cgi?id=290369
rdar://147940452

Reviewed by Anne van Kesteren and Darin Adler.

Implement:
- Add preference
- IDL changes specified at: https://html.spec.whatwg.org/#htmlelement
- DOM getter/setter algorithms specified at: https://html.spec.whatwg.org/#dom-hidden

* LayoutTests/imported/w3c/web-platform-tests/html/editing/the-hidden-attribute/hidden-idl-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/editing/the-hidden-attribute/hidden-until-found-idl-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::hidden const):
(WebCore::HTMLElement::setHidden):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLElement.idl:

Canonical link: https://commits.webkit.org/292735@main
